### PR TITLE
Rustc cache

### PR DESCRIPTION
**Pull request by matklad**
This implements rustc caching, to speed-up no-op builds. 

The cache is per-project, and stored in `target` directory. To implement this, I had to move `rustc` from `Config` down to `BuildConfig`. 

closes https://github.com/rust-lang/cargo/issues/5315

### CODE REVIEW

File: src/cargo/util/rustc.rs [hunk 1]
```diff
@@ -62,4 +75,157 @@ impl Rustc {
...
+            rustup_home.hash(&mut hasher);
+            let rustup_rustc = Path::new(&rustup_home)
+                .join("toolchains")
+                .join(rustup_toolchain)
+                .join("bin")
+                .join("rustc");
+            paths::mtime(&rustup_rustc)?.hash(&mut hasher);
+        }
+        _ => (),
+    }
```
Comment (matklad) [inline src/cargo/util/rustc.rs]: This match requires some scrutiny during review, because it is not tested, but kinda important :-) 

We can't just look at `RUSTUP_TOOLCHAIN`, because the toolchain name does not change when you do `rustup update`, it is just `stable-x86_64-unknown-linux-gnu`. So we actually have to look inside the toolchain and mtime the real rustc binary. 

This still has a failure mode unfortunately :( If the `cargo` is not rustup, but `rustc` is rustup, then these env variables won't be set, and we won't be able to detect compieler changes.

File: src/cargo/util/config.rs [hunk 2]
```diff
@@ -158,13 +158,11 @@ impl Config {
...
 
     /// Get the path to the `rustc` executable
-    pub fn rustc(&self) -> CargoResult<&Rustc> {
-        self.rustc.try_borrow_with(|| {
-            Rustc::new(
-                self.get_tool("rustc")?,
-                self.maybe_get_tool("rustc_wrapper")?,
-            )
-        })
+    pub fn new_rustc(&self) -> CargoResult<Rustc> {
```
Comment (djc) [inline src/cargo/util/config.rs]: Why rename this here? Is that just cosmetic? Maybe extract it into a separate commit that explains why this makes sense on its own?

File: src/cargo/core/compiler/context/compilation_files.rs [hunk 3]
```diff
@@ -434,9 +434,7 @@ fn compute_metadata<'a, 'cfg>(
     unit.target.name().hash(&mut hasher);
     unit.target.kind().hash(&mut hasher);
 
-    if let Ok(rustc) = cx.config.rustc() {
-        rustc.verbose_version.hash(&mut hasher);
-    }
+    cx.build_config.rustc.verbose_version.hash(&mut hasher);
```
Comment (djc) [inline src/cargo/core/compiler/context/compilation_files.rs]: In my perception, it seemed that `build_config` should probably stay immutable over the lifetime of the `Context` (and in fact that is enforced, I think, by #5358).

**Comment by matklad** [comment]
>I don't quite get what the earlier refactorings are trying to achieve. However, I highly appreciate that we should probably cache the output of the rustc invocation. Can't we just hide the cache inside Config::rustc(

I’ve initially tried to just add caching to Config.rustc, but ended up moving rustc to BuildConfig first. The reason is that we need to store the cache somewhere, and the logical choice is to store it in the target directory of the project. However, Config does not know the target directory, because it does not know the current project.

**Comment by djc** [comment]
> I’ve initially tried to just add caching to Config.rustc, but ended up moving rustc to BuildConfig first. The reason is that we need to store the cache somewhere, and the logical choice is to store it in the target directory of the project. However, Config does not know the target directory, because it does not know the current project.
Yeah, that part actually makes sense to me -- I had been thinking about doing the same. However, it seems you've made the mechanics of doing so more complicated than what seems necessary. I guess you had some other goals at the same time, or maybe this `rustc` thing is more complicated than what I've understood so far.

File: src/cargo/util/paths.rs [hunk 4]
```diff
@@ -129,6 +158,12 @@ pub fn append(path: &Path, contents: &[u8]) -> CargoResult<()> {
     Ok(())
 }
 
+pub fn mtime(path: &Path) -> CargoResult<FileTime> {
+    let meta =
+        fs::metadata(path).chain_err(|| internal(format!("failed to stat `{}`", path.display())))?;
```
Comment (alexcrichton) [inline src/cargo/util/paths.rs]: In general `internal` hasn't worked out great for Cargo and I've been trying to remove it wherever I see it, so while you're here mind removing this `internal`?

File: src/cargo/util/rustc.rs [hunk 5]
```diff
@@ -62,4 +75,157 @@ impl Rustc {
...
     }
+
+    pub fn cached_output(&self, cmd: &ProcessBuilder) -> CargoResult<(String, String)> {
+        self.cache.lock().unwrap().cached_output(cmd)
+    }
+}
+
+/// It is a well known that `rustc` is not the fastest compiler in the world.
+/// What is less known is that even `rustc --version --verbose` takes about a
+/// hundred milliseconds! Because we need compiler version info even for no-op
```
Comment (alexcrichton) [inline src/cargo/util/rustc.rs]: Could this cc the Cargo/rustc issues as well?

File: src/cargo/util/rustc.rs [hunk 6]
```diff
@@ -62,4 +75,157 @@ impl Rustc {
...
+                self.dirty = true;
+                Ok(output)
+            }
+        }
+    }
+}
+
+impl Drop for Cache {
+    fn drop(&mut self) {
+        match (&self.cache_location, self.dirty) {
```
Comment (alexcrichton) [inline src/cargo/util/rustc.rs]: Perhaps this could be started with:

```rust
if !self.dirty {
    return 
}
```

Comment (alexcrichton) [reply src/cargo/util/rustc.rs]: Er well I'll leave it up to you, I'm just personally always looking to aggressively reduce indent!

File: src/cargo/util/rustc.rs [hunk 7]
```diff
@@ -62,4 +75,157 @@ impl Rustc {
...
+
+fn rustc_fingerprint(path: &Path) -> CargoResult<u64> {
+    let mut hasher = SipHasher::new_with_keys(0, 0);
+
+    let path = paths::resolve_executable(path)?;
+    path.hash(&mut hasher);
+
+    paths::mtime(&path)?.hash(&mut hasher);
+
+    match (env::var("RUSTUP_HOME"), env::var("RUSTUP_TOOLCHAIN")) {
```
Comment (alexcrichton) [inline src/cargo/util/rustc.rs]: Could you insert a comment here as to what's going on?

File: src/cargo/util/rustc.rs (hunk 1 above)
Comment (alexcrichton) [reply src/cargo/util/rustc.rs]: One possible mitigation here I think would be to look at the result of the resolved executable. If `rustc` resolves to `$CARGO_HOME/.bin/rustc` then I think we can say with a high level of certainty that it's a rustup rustc. We could even say that if we detect a rustup rustc and `RUSTUP_HOME` and/or `RUSTUP_TOOLCHAIN` isn't set that we skip this cache entirely.

File: src/cargo/ops/cargo_compile.rs [hunk 8]
```diff
@@ -260,7 +260,8 @@ pub fn compile_ws<'a>(
         bail!("jobs must be at least 1")
     }
 
-    let mut build_config = BuildConfig::new(config, jobs, &target)?;
+    let rustc_info_cache = ws.target_dir().join("rustc_info.json").into_path_unlocked();
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_compile.rs]: Mind throwing a `.` on the front to hide it away?

**Comment by matklad** [comment]
>However, it seems you've made the mechanics of doing so more complicated than what seems necessary. 

Quite probably, but, if I did this, it is because of being stupid, and not because of being smart :-) What simplifications would you suggest?

File: src/cargo/util/rustc.rs (hunk 1 above)
Comment (matklad) [reply src/cargo/util/rustc.rs]: Yep, implemented that logic!

File: src/cargo/util/rustc.rs (hunk 5 above)
Comment (matklad) [reply src/cargo/util/rustc.rs]: Done!

File: src/cargo/util/rustc.rs (hunk 6 above)
Comment (matklad) [reply src/cargo/util/rustc.rs]: Yeah, it looks much nicer that way, because I don't have to wrangle with not moving out cahce_location as well!

File: src/cargo/util/rustc.rs (hunk 7 above)
Comment (matklad) [reply src/cargo/util/rustc.rs]: Done!

File: src/cargo/ops/cargo_compile.rs (hunk 8 above)
Comment (matklad) [reply src/cargo/ops/cargo_compile.rs]: :heavy_check_mark:

File: src/cargo/core/compiler/context/compilation_files.rs (hunk 3 above)
Comment (matklad) [reply src/cargo/core/compiler/context/compilation_files.rs]: Hm, not sure I understand this comment :) `build_config` is not mutated here at all?

**Comment by matklad** [comment]
For posterity, a couple of more details:

* write to the `.rustc_info.json` is racy across Cargo processes. This shoudn't be a problem because we always ignore errors in this file and, in practice, it should be mostly read-only because we write cache only if something did change. 

* it's interesting to compare this with what `sccache` does. `sccache` is actually pretty smart: for compiler fingerprint, it hashes the contents of `rustc --print sysroot`, which looks like a bulletproof way to do compiler fingerprinting. Unfortunately, we can't use this here because it requires to launch compiler once, the very thing we'd love to avoid.

**Comment by matklad** [comment]
Let's tag it relnotes, because some folks might need to be aware of `CARGO_CACHE_RUSTC_INFO`, if they use something like rustup, but not rustup, to manage toolchains.

**Comment by luser** [comment]
> it's interesting to compare this with what sccache does. sccache is actually pretty smart: for compiler fingerprint, it hashes the contents of rustc --print sysroot, which looks like a bulletproof way to do compiler fingerprinting. Unfortunately, we can't use this here because it requires to launch compiler once, the very thing we'd love to avoid.

Coming in here late but I ran into this issue while reading old release notes. Note that sccache only uses the hash of the sysroot libs as input to the hash function for compilation caching--it doesn't recalculate that hash every time (that'd be expensive!) it caches that hash internally with the same strategy you're using here (by comparing mtimes). sccache doesn't have the checks for rustup you added, we've got [an open issue](https://github.com/mozilla/sccache/issues/87) on handling rustup properly.
